pkg/observer/controller.go

```diff
@@ -230,21 +230,42 @@ func (c *controller) inProgressCNRs() v1.CycleNodeRequestList {
 
 // dropInProgressNodeGroups matches nodeGroups to CNRs and filters out any that match
 func (c *controller) dropInProgressNodeGroups(nodeGroups v1.NodeGroupList, cnrs v1.CycleNodeRequestList) v1.NodeGroupList {
-	// filter out nodegroups that aren't current in progress with a cnr
+	// Filter out nodegroups that aren't currently in progress with a cnr. Count
+	// failed CNRs only if they don't outnumber the max threshold defined for
+	// the nodegroup
 	var restingNodeGroups v1.NodeGroupList
+
 	for i, nodeGroup := range nodeGroups.Items {
-		var found bool
+		var dropNodeGroup bool
+		var failedCNRsFound uint
+
 		for _, cnr := range cnrs.Items {
-			if sameNodeGroups(cnr.GetNodeGroupNames(), nodeGroup.GetNodeGroupNames()) {
-				found = true
-				break
+			// CNR doesn't match nodegroup, skip it
+			if !cnr.IsFromNodeGroup(nodeGroup) {
+				continue
+			}
+
+			// Count the Failed CNRs separately, they need to be counted before
+			// they can be considered to drop the nodegroup
+			if cnr.Status.Phase == v1.CycleNodeRequestFailed {
+				failedCNRsFound++
+			} else {
+				dropNodeGroup = true
+			}
+
+			// If the number of Failed CNRs exceeds the threshold in the
+			// nodegroup then drop it
+			if failedCNRsFound > nodeGroup.Spec.MaxFailedCycleNodeRequests {
+				dropNodeGroup = true
 			}
 		}
-		if found {
+
+		if dropNodeGroup {
 			klog.Warningf("nodegroup %q has an in progress CNR.. skipping this nodegroup", nodeGroup.Name)
```

Comment:
This log message now doesn't match reality. dropNodeGroup is being set to true when the number of failed CNRs exceeds the threshold, and not when it has an in progress CNR.

Reply:
I will add a second log line there because `Failed` is defined as "in progress" in the existing implementation. I'm adding functionality to skip a certain number of them from counting as "in progress" so I will update for that.